### PR TITLE
[inductor] Add a subprocess-based parallel compile

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126817
* __->__ #126816

Summary:
Adds a "safe" parallel compile implementation that a) Popens a sub-process with an entry point we control, and b) Uses a ProcessPoolExecutor in that sub-processes to perform parallel compiles. This change essentially squashes these two implementations from jansel, but removes the "thread-based" approach since benchmarking revealed that compile-time performance was poor compared to the existing impl:
https://github.com/pytorch/pytorch/pull/124682
https://github.com/pytorch/pytorch/pull/122941

This PR adds the implementation, but defaults to the existing "fork". I'll submit a separate change to enable.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by masnesral** [comment]
Benchmark results: https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C+14+May+2024+19%3A50%3A25+GMT&stopTime=Tue%2C+21+May+2024+19%3A50%3A25+GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh%2Fmasnesral%2F61%2Fhead&lCommit=e5e2e5796b0ffab0922da6a7a714a26ae57ba513&rBranch=main&rCommit=53f73cdeb6fe53155f91064ca15b722e80dec2f3

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -3177,6 +3137,8 @@ def wait(self, scope: Dict[str, Any]) -> None:
 if (
     os.environ.get("TORCH_TNT_IN_USE", "0") == "1"
     or os.environ.get("TORCH_WARM_POOL", "1") != "1"
+    # TODO(masnesral): warm pool is broken, need to fix it
```
Comment (masnesral) [inline torch/_inductor/codecache.py:3140]: @jansel, I commandeered this TODO you had originally, but I'm not sure there's anything to actually do here? We're calling _warm_process_pool() on the ProcessPoolExecutor owned by the subproc in SubprocMain.__init__(). Or did you have some other kind of warmup?

Comment (jansel) [reply torch/_inductor/codecache.py:3140]: I was getting errors with this turned on.  Does it pass for you?

**Review by jansel — APPROVED** [review]

File: torch/_inductor/compile_worker/__main__.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import argparse
+import os
+import sys
+import typing
+
+from torch._inductor.codecache import _set_triton_ptxas_path, caching_device_properties
```
Comment (jamesjwu) [inline torch/_inductor/compile_worker/__main__.py:6]: `_set_triton_ptxas_path` isn't defined in torch._inductor.codecache, do you mean the one defined in compile_tasks.py?

Comment (masnesral) [reply torch/_inductor/compile_worker/__main__.py:6]: Hmmm; I guess it works because codecache.py imports that function. But yeah, will fix

File: torch/_inductor/compile_worker/subproc_pool.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    # Examples:
+    # A simple x + x + x script: 10ms seconds in the middle of the program, 2ms at startup
+    # tf_efficientnet_b0 benchmark: 50ms! in the middle of the program , 3ms at startup
+
+    # So we want to start the workers early when it is still cheap, and also to allow the workers to get
+    # ready before we have work for them.
+
+    # ProcessPoolExecutor also does not launch the workers until it finds a point when all the workers are idle.
+    # But if we waited until then fork time will be long and we will be waiting for the processes to initialize.
```
Comment (eellison) [inline torch/_inductor/compile_worker/subproc_pool.py:237]: Are these comments still relevant ?

File: torch/_inductor/compile_worker/__main__.py (hunk 2 above)
Comment (jamesjwu) [reply torch/_inductor/compile_worker/__main__.py:6]: python torch/_inductor/compile_worker/__main__.py doesn't seem to run for me if you just call it directly, not sure if there's some special way to call it so it passes today

Comment (masnesral) [reply torch/_inductor/compile_worker/__main__.py:6]: I'm perplexed. Seems to work fine for me:
```
$ python torch/_inductor/compile_worker/__main__.py --help
usage: __main__.py [-h] [--workers WORKERS] [--parent PARENT]

options:
  -h, --help         show this help message and exit
  --workers WORKERS
  --parent PARENT
```

(but making change anyway)

File: torch/_inductor/compile_worker/subproc_pool.py (hunk 3 above)
Comment (masnesral) [reply torch/_inductor/compile_worker/subproc_pool.py:237]: Lemme add a TODO and I'll benchmark

**Comment by masnesral** [comment]
New benchmark result after properly warming the cache:

https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2021%20May%202024%2003%3A36%3A49%20GMT&stopTime=Tue%2C%2028%20May%202024%2003%3A36%3A49%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/masnesral/65/head&lCommit=4293f564e1bbd0724578cecc05e6159b4166ce93&rBranch=main&rCommit=b516de8cac9317155104d476056bae1d910a125c
